acme/acme/standalone.py

```diff
@@ -0,0 +1,230 @@
+
+    :ivar set simple_http_resources: A set of `SimpleHTTPResource`
+        objects. TODO: better name?
+
+    """
+    SimpleHTTPResource = collections.namedtuple(
+        "SimpleHTTPResource", "chall response validation")
+
+    def __init__(self, *args, **kwargs):
+        self.simple_http_resources = kwargs.pop("simple_http_resources", set())
+        socketserver.BaseRequestHandler.__init__(self, *args, **kwargs)
+
+    def do_GET(self):  # pylint: disable=invalid-name,missing-docstring
+        if self.path == "/":
+            self.handle_index()
+        elif self.path.startswith("/" + challenges.SimpleHTTP.URI_ROOT_PATH):
+            self.handle_simple_http_resource()
+        else:
+            self.handle_404()
+
+    def handle_index(self):
+        """Handle index page."""
+        self.send_response(200)
+        self.send_header("Content-type", "text/html")
+        self.end_headers()
+        self.wfile.write(self.server.server_version)
+
+    def handle_404(self):
+        """Handler 404 Not Found errors."""
+        self.send_response(http_client.NOT_FOUND, message="Not Found")
+        self.send_header("Content-type", "text/html")
+        self.end_headers()
+        self.wfile.write("404")
+
+    def handle_simple_http_resource(self):
+        """Handle SimpleHTTP provisioned resources."""
+        for resource in self.simple_http_resources:
+            if resource.chall.path == self.path:
+                logger.debug("Serving SimpleHTTP with token %r",
+                             resource.chall.encode("token"))
+                self.send_response(http_client.OK)
+                self.send_header("Content-type", resource.response.CONTENT_TYPE)
+                self.end_headers()
+                self.wfile.write(resource.validation.json_dumps().encode())
+                return
+        else:  # pylint: disable=useless-else-on-loop
+            logger.debug("No resources to serve")
+        logger.debug("%s does not correspond to any resource. ignoring",
+                     self.path)
+
+    @classmethod
+    def partial_init(cls, simple_http_resources):
+        """Partially initialize this handler.
+
+        This is useful because `socketserver.BaseServer` takes
+        uninitialized handler and initializes it with the current
+        request.
+
+        """
+        return functools.partial(
+            cls, simple_http_resources=simple_http_resources)
+
+
+class ACMERequestHandler(SimpleHTTPRequestHandler):
+    """ACME request handler."""
+
+    def handle_one_request(self):
+        """Handle single request.
+
+        Makes sure that DVSNI probers are ignored.
+
+        """
+        try:
+            return SimpleHTTPRequestHandler.handle_one_request(self)
+        except OpenSSL.SSL.ZeroReturnError:
+            logger.debug("Client prematurely closed connection (prober?). "
+                         "Ignoring request.")
+
+
+def simple_server(cli_args, forever=True):
```

Comment:
`acme` is meant to be reusable library, generic implementation of the ACME protocol. This bit of code allows experimenters to simply `python -m acme.standalone`, just like you can do it with `python -m SimpleHTTPServer` from stdlib (see `acme/examples/standalone`) - quick way to solve challenges.